openedx/custom/db/fields.py

```diff
@@ -0,0 +1,77 @@
+
+
+class MultiSelectWithOtherField(MultiSelectField):
+
+    def __init__(self,other_max_length=None,*args, **kwargs):
+        self.other_max_length = other_max_length
+
+        if kwargs.get('max_length') is None and self.other_max_length is not None:
+            choice_max_length = get_max_length(kwargs['choices'], kwargs.get('max_length'))
+            kwargs['max_length'] = choice_max_length + self.other_max_length
+
+        if type(kwargs['choices']) == tuple:
+            kwargs['choices'] = kwargs.pop('choices') + (('other', 'Other'),)
+        super(MultiSelectWithOtherField, self).__init__(*args, **kwargs)
+        self.error_messages.update({
+            'invalid_char': 'value %s contains invalid character `|`'
+        })
+
+    def get_prep_value(self, value):
+        selected_value = other_value = ''
+        choice_values = [choice[0] for choice in self.choices]
+        if value:
+            for val in value:
+                if val in choice_values:
+                    selected_value += val + ','
+                else:
+                    other_value = val
+
+            selected_value += '|' + other_value
+        return selected_value
+
+    def formfield(self, **kwargs):
+        defaults = {
+            'required': not self.blank,
+            'label': capfirst(self.verbose_name),
+            'help_text': self.help_text,
+            'choices': self.choices,
+            'max_length': self.max_length,
+            'max_choices': self.max_choices,
+            'other_max_length': self.other_max_length
+        }
+        if self.has_default():
+            defaults['initial'] = self.get_default()
+        defaults.update(kwargs)
+        return MultiSelectWithOtherFormField(**defaults)
+
+    def validate(self, value, model_instance):
+        """
+        this function is to validate the input values for multi select field, however we are implementing field with
+        support of other input filed we are disabling validations to let other input text pass to the database.
+        :param value: list of all selected choice with other text.
+        :param model_instance: current model instance for with it is saving data.
+        :return: nothing
+        """
+        for opt_select in value:
+            if '|' in opt_select:
+                raise exceptions.ValidationError(self.error_messages['invalid_char'] % value)
+
+    def to_python(self, value):
+        choices = dict(self.flatchoices)
+        if value:
+            if isinstance(value, list):
+                return value
+            elif isinstance(value, string_type):
+                choices_str = value.replace('|', '')
+                selected_choices = [choice for choice in choices_str.split(',') if choice.strip()]
+                return MSFList(choices, selected_choices)
+            elif isinstance(value, (set, dict)):
+                return MSFList(choices, list(value))
+        return MSFList(choices, [])
```

Comment:
@harris-irfan the style you are suggesting is apparently giving me the feel we writing it with the intention to execute multiple conditional blocks, as it will full fill the purpose but from the reader's point of view, it will require some more cycles of readers brain :P

Reply:
@qasimgulzar that makes sense. Thank you for pointing this out. But can we at least follow the early exit rule to avoid nesting:

`if not value:
    return MSFList(choices, [])
if isinstance(value, list):
    return value
elif isinstance(value, string_type):
    choices_str = value.replace('|', '')
    selected_choices = [choice for choice in choices_str.split(',') if choice.strip()]
    return MSFList(choices, selected_choices)
elif isinstance(value, (set, dict)):
    return MSFList(choices, list(value))

`

Reply:
@harris-irfan it is totally opinion based coding style, another there is also a possibility that none of the nested if will be passed so, in that case, I will again have to place return on the last line.